Rework asset bundle loading for TLD 2.06 (catalogs & addressables)

dommrogers opened the pull request:
https://github.com/dommrogers/ModComponent/blob/master/docs/Convert-Bundle.md
for convert process.

dommrogers committed:
WIP - TLD 2.06

very close.... It builds & loads

dommrogers committed:
remove salt & yeast

dommrogers committed:
change Resources.Load to GearItem.LoadGearItemPrefab for base game items

dommrogers committed:
comment out "_dmg_texture" elements

player.log reports it doesn't exist.

dommrogers committed:
fix console search/autocomplete for added items

dommrogers committed:
add using Il2CppTLD.Gear;

dommrogers committed:
almost there....

dommrogers committed:
Il2CppAK.Wwise.Event emptyEvent

dommrogers committed:
comment out AssetReferenceGearItem (unsure if required)

note: I added this in, was not previously set

dommrogers committed:
remove some debug output

dommrogers committed:
use UTF8 to read files

dommrogers committed:
add AssetReferenceGearItem & guid to GearItem

dommrogers committed:
ResourcePatches cleanup & start fresh (Resources.Load x 2 Done)

dommrogers committed:
avoid using Dict.TryGetValue

dommrogers committed:
WIP patch asset & texture loading

dommrogers committed:
WIP

dommrogers committed:
fix cooking audio for cookable items

dommrogers committed:
use new Json method name for LocalizationUtilities

dommrogers committed:
remove incorrect m_DisplayNameOverrideLocID

dommrogers committed:
functional with new .mc packs (needs cleanup)

dommrogers committed:
push convert images

dommrogers committed:
missing image

dommrogers committed:
convert-bundle.md

dommrogers committed:
update convert-bundle.md

dommrogers committed:
update shader_naming.png

dommrogers committed:
cleanup convert-bundle.md

dommrogers committed:
improve catalog loading order & add checks

dommrogers committed:
Little cleanup for PR

dommrogers committed:
adjust convert-bundle.md

dommrogers committed:
add CI & NuGet

dommrogers committed:
cleanup old #warning

dommrogers committed:
add more catalog checking

dommrogers committed:
version bump 6.0.0

dommrogers committed:
fix CI artifact names :)

ds5678 commented on ModComponent/API/Components/ModPowderComponent.cs:
```diff
@@ -44,20 +44,13 @@ internal static GearPowderType GetPowderType(ModPowderType modPowderType)
-				? ModPowderType.Yeast 
-				: ModPowderType.Gunpowder;
+		return ModPowderType.Gunpowder;
 	}
```
Delete both these methods.

ds5678 commented on ModComponent/API/Components/ModRandomItemComponent.cs:
```diff
@@ -1,9 +1,10 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove line

ds5678 commented on ModComponent/Utils/AssetBundleUtils.cs:
```diff
@@ -45,4 +49,16 @@ public static UnityEngine.Object LoadAsset(AssetBundle assetBundle, string name,
 		}
 		return assetBundle.LoadAsset_Internal(name, type);
 	}
+
+	
+
+	
+
+
+
+
+
+
+
+
```
Delete

ds5678 commented on ModComponent/API/Behaviours/ModMillableBehaviour.cs:
```diff
@@ -1,6 +1,6 @@
 ﻿using MelonLoader.TinyJSON;
 using System;
-using UnhollowerBaseLib.Attributes;
+using Il2CppInterop.Runtime.Attributes;
 using UnityEngine;
```
Visual Studio has a cleanup tool to remove and sort usings on the whole project. Do that.

ds5678 commented on ModComponent/API/Behaviours/ModStackableBehaviour.cs:
```diff
@@ -1,8 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove space

ds5678 commented on ModComponent/API/Components/ModCookableComponent.cs:
```diff
@@ -81,7 +84,10 @@ internal override void InitializeComponent(ProxyObject dict, string className =
 		{
-			this.CookingResult = Resources.Load(dict.GetVariant(className, "CookingResult")).Cast<GameObject>();
+			string assetName = dict.GetVariant(className, "CookingResult").ToString();
+			//MelonLoader.MelonLogger.Warning("Cookable - " + assetName);
```
Remove

ds5678 commented on ModComponent/API/Components/ModCookableComponent.cs:
```diff
@@ -81,7 +84,10 @@ internal override void InitializeComponent(ProxyObject dict, string className =
+			string assetName = dict.GetVariant(className, "CookingResult").ToString();
+			//MelonLoader.MelonLogger.Warning("Cookable - " + assetName);
+			this.CookingResult = Addressables.LoadAssetAsync<GameObject>(assetName).WaitForCompletion();
+			//MelonLoader.MelonLogger.Warning("Cookable - " + this.CookingResult.ToString());
```
Remove

ds5678 commented on ModComponent/API/Components/ModCookableComponent.cs:
```diff
@@ -81,7 +84,10 @@ internal override void InitializeComponent(ProxyObject dict, string className =
-			this.CookingResult = Resources.Load(dict.GetVariant(className, "CookingResult")).Cast<GameObject>();
+			string assetName = dict.GetVariant(className, "CookingResult").ToString();
+			//MelonLoader.MelonLogger.Warning("Cookable - " + assetName);
+			this.CookingResult = Addressables.LoadAssetAsync<GameObject>(assetName).WaitForCompletion();
```
In AssetBundleUtils, add a method for loading assets from addressables.

```cs
public static T LoadAsset<T>(string assetName) => Addressables.LoadAssetAsync<T>(assetName).WaitForCompletion();
```

ds5678 commented on ModComponent/API/Components/ModExplosiveComponent.cs:
```diff
@@ -1,7 +1,7 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/API/Components/ModPowderComponent.cs:
```diff
@@ -1,16 +1,16 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/API/Components/ModPowderComponent.ModPowderType.cs:
```diff
@@ -1,13 +1,11 @@
 	public enum ModPowderType
 	{
-		Gunpowder,
-		Salt,
-		Yeast
+		Gunpowder
 	}
```
The Hinterland enums have been shown to be stable over large time periods. I think I want most of the "mod" enums removed. Any "mod" enums that aren't direct copies can stay.

ds5678 commented on ModComponent/API/Components/ModRandomItemComponent.cs:
```diff
@@ -36,7 +37,7 @@ void Update()
 
 		int index = RandomUtils.Range(0, this.ItemNames.Length);
-		GameObject? prefab = Resources.Load(this.ItemNames[index])?.Cast<GameObject>();
+		GameObject? prefab = Addressables.LoadAssetAsync<GameObject>(this.ItemNames[index]).WaitForCompletion()?.Cast<GameObject>();
```
`?.Cast<GameObject>()` is unnecessary.

ds5678 commented on ModComponent/API/Components/ModRandomWeightedItemComponent.cs:
```diff
@@ -1,10 +1,11 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/API/Components/ModRandomWeightedItemComponent.cs:
```diff
@@ -54,7 +55,7 @@ void Update()
 
 		int index = this.GetIndex();
-		GameObject? prefab = Resources.Load(this.ItemNames[index])?.Cast<GameObject>();
+		GameObject? prefab = Addressables.LoadAssetAsync<GameObject>(this.ItemNames[index]).WaitForCompletion()?.Cast<GameObject>();
```
`?.Cast<GameObject>()` is unnecessary.

ds5678 commented on ModComponent/API/ModLootTable.cs:
```diff
@@ -18,11 +18,11 @@ void Awake()
 		for (int i = 0; i < len; i++)
 		{
-			GameObject? go = Resources.Load(prefabNames[i])?.TryCast<GameObject>();
+			GameObject? go = Addressables.LoadAssetAsync<GameObject>(prefabNames[i]).WaitForCompletion()?.TryCast<GameObject>();
```
TryCast is unnecessary

ds5678 commented on ModComponent/AssetLoader/AtlasUtils.cs:
```diff
@@ -1,5 +1,5 @@
-using Hinterland;
+﻿using Il2Cpp;
+
 
```
Extra space

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -27,7 +24,7 @@ internal static class ModAssetBundleManager
 	/// Value: The cooresponding loaded asset bundle
 	/// </summary>
-	private static readonly Dictionary<string, AssetBundle> knownAssetBundles = new Dictionary<string, AssetBundle>();
+	internal static readonly Dictionary<string, AssetBundle> knownAssetBundles = new Dictionary<string, AssetBundle>();
```
Don't break encapsulation. Find another way.

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -37,7 +34,7 @@ internal static class ModAssetBundleManager
 	/// Value: The asset bundle containing the asset
 	/// </summary>
-	private static readonly Dictionary<string, AssetBundle> knownAssetNames = new Dictionary<string, AssetBundle>();
+	internal static readonly Dictionary<string, AssetBundle> knownAssetNames = new Dictionary<string, AssetBundle>();
```
Don't break encapsulation. Find another way.

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -193,10 +190,12 @@ public static string GetFullAssetName(string? name)
 		string lowerCaseName = name!.ToLowerInvariant();
 		if (knownAssetNames.ContainsKey(lowerCaseName))
 		{
+			MelonLoader.MelonLogger.Warning("knownAssetNames " + lowerCaseName);
```
Remove

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -193,10 +190,12 @@ public static string GetFullAssetName(string? name)
 		string lowerCaseName = name!.ToLowerInvariant();
```
Remove !

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -281,9 +280,10 @@ private static void LoadAssetBundle(string relativePath, AssetBundle assetBundle
 		}
 
-		Logger.Log(stringBuilder.ToString().Trim());
+		//Logger.Log(stringBuilder.ToString().Trim());
```
Undo

ds5678 reviewed (changes requested):
Some starting requests

dommrogers committed:
remove ResourcePatches, no longer required with addressables

dommrogers committed:
PR Changes #1

dommrogers committed:
PR Changes #1.1

Removed all 1:1 ModEums
Added LoadAsset wrapper method

ds5678 commented on ModComponent/API/Components/ModToolComponent.cs:
```diff
@@ -1,6 +1,8 @@
+using MelonLoader.TinyJSON;
 using ModComponent.Utils;
-using UnhollowerBaseLib.Attributes;
+
```
Remove

ds5678 commented on ModComponent/API/Modifications/PlayAkSound.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/AssetLoader/AtlasUtils.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/AssetLoader/ModAssetBundleManager.cs:
```diff
@@ -281,6 +278,7 @@ private static void LoadAssetBundle(string relativePath, AssetBundle assetBundle
 			stringBuilder.Append(" => ");
 			stringBuilder.Append(eachAssetName);
 			stringBuilder.Append("\n");
+
```
Remove

ds5678 commented on ModComponent/Mapper/AlternativeToolManager.cs:
```diff
@@ -1,7 +1,7 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/AccelerantMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/BurnableMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/CarryingCapacityMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/EvolveMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/FireStarterMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/HarvestableMapper.cs:
```diff
@@ -1,9 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/MillableMapper.cs:
```diff
@@ -1,9 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/RepairableMapper.cs:
```diff
@@ -1,9 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/ScentMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/SharpenableMapper.cs:
```diff
@@ -1,9 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/StackableMapper.cs:
```diff
@@ -1,9 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/BehaviourMappers/TinderMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/BedMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/BedMapper.cs:
```diff
@@ -18,7 +18,7 @@ internal static void Configure(ModBaseComponent modComponent)
 		Bed bed = ComponentUtils.GetOrCreateComponent<Bed>(modBedComponent);
 
-		bed.m_LocalizedDisplayName = NameUtils.CreateLocalizedString(modComponent.DisplayNameLocalizationId);
+		bed.name = NameUtils.CreateLocalizedString(modComponent.DisplayNameLocalizationId).Text();
```
I feel like there's a cleaner way to do this. `Localization.GetString` should be better.

However, both of those feel wrong because they depend on the current localization, which can be changed in-game.

dommrogers replied:
yeah `m_LocalizedDisplayName` no longer exists on `Bed` component so I wasn't sure either.

ds5678 replied:
My intuition says that this name property changes the name of the GameObject it's attached to.

dommrogers replied:
removed completely, tested with an updated CampingCot MC bundle, hover text comes from the gearItem now.

![image](https://user-images.githubusercontent.com/707295/227248519-cc22d9e9-2299-4a0d-8d66-ea6b9737577a.png)

![image](https://user-images.githubusercontent.com/707295/227248709-4f711e0d-54fa-4cba-a3a2-1f69f5a5af3c.png)

ds5678 commented on ModComponent/Mapper/ComponentMappers/BodyHarvestMapper.cs:
```diff
@@ -1,6 +1,7 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/CharcoalMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/ClothingMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/CollectibleMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/CookingPotMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/FirstAidMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/FoodMapper.cs:
```diff
@@ -1,7 +1,8 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/InspectMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/LiquidMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/PowderMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/PowderMapper.cs:
```diff
@@ -16,7 +16,7 @@ internal static void Configure(ModBaseComponent modComponent)
 
 		PowderItem powderItem = ComponentUtils.GetOrCreateComponent<PowderItem>(modComponent);
-		powderItem.m_Type = ModPowderComponent.GetPowderType(modPowderComponent.PowderType);
+		powderItem.m_Type = GearPowderType.Gunpowder;
```
`modPowderComponent.PowderType`

ds5678 commented on ModComponent/Mapper/ComponentMappers/PurificationMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/ResearchMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ComponentMappers/ToolMapper.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ConsoleWaitlist.cs:
```diff
@@ -59,10 +102,12 @@ internal struct ModConsoleName
 	{
```
internal readonly struct

ds5678 commented on ModComponent/Mapper/GearEquipper.cs:
```diff
@@ -1,6 +1,7 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Mapper/ItemMapper.cs:
```diff
@@ -116,91 +115,110 @@ private static void ConfigureGearItem(ModBaseComponent modComponent)
 		gearItem.m_SkinnedMeshRenderers = ModUtils.NotNull<SkinnedMeshRenderer>(gearItem.m_SkinnedMeshRenderers);
 
-		GameObject template = Resources.Load<GameObject>("GEAR_CoffeeCup");
+		
```
Remove

ds5678 commented on ModComponent/Mapper/ItemMapper.cs:
```diff
@@ -211,17 +229,13 @@ private static void PostProcess(ModBaseComponent modComponent)
 				{
 					eachMaterial.shader = meshRenderer.material.shader;
 					eachMaterial.shaderKeywords = meshRenderer.material.shaderKeywords;
-
-					if (eachMaterial.GetTexture("_dmg_texture") == null)
-					{
-						eachMaterial.SetTexture("_dmg_texture", eachMaterial.GetTexture("_MainTex"));
-					}
```
This is important and cannot be removed.

dommrogers replied:
This is why I removed it.
```
Material 'rahatLokum (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'roastedAlmonds (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'solitudeCider (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'strawberryPlombir (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'tunaPate (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'Standard (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'
Material 'Standard (Instance)' with Shader 'Shader Forge/TLD_StandardSkinned' doesn't have a texture property '_dmg_texture'```

ds5678 replied:
Well yes, food doesn't need a damage property. This is for clothing.

dommrogers replied:
restored.

ds5678 commented on ModComponent/Mapper/ZipFileLoader.cs:
```diff
@@ -14,7 +11,7 @@ namespace ModComponent.Mapper;
 {
 	internal static readonly List<byte[]> hashes = new();
-
+	
```
Undo

ds5678 commented on ModComponent/Mapper/ZipFileLoader.cs:
```diff
@@ -84,17 +81,9 @@ private static void LoadZipFile(string zipFilePath)
 	{
-		Encoding encoding = GetEncoding(memoryStream);
-		return encoding.GetString(memoryStream.ToArray());
+		return Encoding.UTF8.GetString(memoryStream.ToArray());
```
I think this can directly use the stream without converting to an array.

dommrogers replied:
`Argument 1: cannot convert from 'System.IO.MemoryStream' to 'byte[]'`

ds5678 replied:
`return new StreamReader(memoryStream, Encoding.UTF8).ReadToEnd();`

ds5678 replied:
You may have to check the first few bytes for BOM and possibly use a cached `new Utf8Encoding(false)`.

dommrogers replied:
Implemented same BOM check as LocalizationUtils

ds5678 replied:
That was fine in LocalizationUtilities because the file would always be json, but not here. I want a more general purpose solution.
```cs
	private static string ReadToString(MemoryStream memoryStream)
	{
		ReadOnlySpan<byte> utf8BOM = stackalloc byte[] { 0xEF, 0xBB, 0xBF };
		ReadOnlySpan<byte> streamData = memoryStream.ToArray();
		ReadOnlySpan<byte> textData = streamData.StartsWith(utf8BOM) ? streamData.Slice(3) : streamData;
		return Encoding.UTF8.GetString(textData);
	}
```

ds5678 commented on ModComponent/Mapper/ZipFileLoader.cs:
```diff
@@ -198,12 +189,19 @@ private static bool TryHandleJson(string zipFilePath, string internalPath, strin
 			{
-				Logger.Log($"Reading json localization from zip at '{internalPath}'");
-				LocalizationUtilities.LocalizationManager.LoadJSONLocalization(text);
+				// change emthod to ensure we go via the BOM fixed methods..
+				TextAsset textAsset = new(text);
+				LocalizationUtilities.LocalizationManager.LoadLocalization(textAsset, internalPath);
```
Changing to TextAsset is not an acceptable solution. Find another way.

dommrogers replied:
Can you elaborate why this isn't acceptable ?

ds5678 replied:
The only reason TextAssets are accepted in LoadLocalization is to reduce friction with loading text files from asset bundles.

Also, `text` should not have BOM at this point. If it does, that is a huge problem.

dommrogers replied:
Restored & Implemented same BOM check as LocalizationUtils

ds5678 commented on ModComponent/Patches/AlternativePowderPatches.cs:
```diff
@@ -10,57 +8,23 @@ internal static class AlternativePowderPatches
 	[HarmonyPatch(typeof(PlayerManager), nameof(PlayerManager.AddPowderToInventory))]
 	private static class PlayerManager_AddPowderToInventory
```
Remove this whole file.

ds5678 commented on ModComponent/Patches/FirestartingPatches.cs:
```diff
@@ -1,19 +1,18 @@
-		{
-			return;
-		}
+		//if (!__result)
+		//{
+		//	return;
+		//}
```
Undo. It's important that the item isn't destroyed when firemaking is unsuccessful.

dommrogers replied:
`public void PlayerStartFire(Fire fire, FireStarterItem starter, FuelSourceItem tinder, FuelSourceItem fuel, FireStarterItem accelerant, Action<bool> onDoneStartingFire)`

Method no longer returns a bool, it is void... Thoughts ?

ds5678 replied:
Is `onDoneStartingFire` a new parameter?

dommrogers replied:
Looks like the method has changed quite a bit.

2.02
`public bool PlayerStartFire(Campfire campfire, FireStarterItem starter, FuelSourceItem tinder, FuelSourceItem fuel, FireStarterItem accelerant, float percentChanceSuccess)`

2.08
`public void PlayerStartFire(Fire fire, FireStarterItem starter, FuelSourceItem tinder, FuelSourceItem fuel, FireStarterItem accelerant, Action<bool> onDoneStartingFire)`

The description for `RuinedAfterUse` is `Set the condition to 0% after the fire starting finished (either successful or not).`
So shouldn't it break regardless of fire success ?

In which case the `if (!__result)` chek was invalid to begin with.

ds5678 commented on ModComponent/Patches/LiquidRandomizationPatches.cs:
```diff
@@ -1,6 +1,6 @@
-using Hinterland;
+﻿using HarmonyLib;
+using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Utils/FileUtils.cs:
```diff
@@ -1,18 +1,12 @@
 
 public static class FileUtils
 {
 	public static string GetModsFolderPath()
```
Remove and inline.

ds5678 commented on ModComponent/Utils/FileUtils.cs:
```diff
@@ -1,18 +1,12 @@
 
 namespace ModComponent.Utils;
 
 public static class FileUtils
```
Make all Utils classes internal

ds5678 commented on ModComponent/Utils/ModContentCatalogData.cs:
```diff
@@ -0,0 +1,30 @@
+﻿namespace ModComponent.Utils;
+
+public class ModContentCatalog
+{
+	public string? m_LocatorId { get; set; }
+	public ModObjectInitializationData? m_InstanceProviderData { get; set; }
+	public ModObjectInitializationData? m_SceneProviderData { get; set; }
+	public List<ModObjectInitializationData>? m_ResourceProviderData { get; set; }
+	public string[]? m_ProviderIds { get; set; }
+	public string[]? m_InternalIds { get; set; }
+	public string? m_KeyDataString { get; set; }
+	public string? m_BucketDataString { get; set; }
+	public string? m_EntryDataString { get; set; }
+	public string? m_ExtraDataString { get; set; }
+	public ModSerializedType[]? m_resourceTypes { get; set; }
+	public string[]? m_InternalIdPrefixes { get; set; }
+}
+
+public class ModObjectInitializationData
+{
+	public string? m_Id { get; set; }
+	public ModSerializedType? m_ObjectType { get; set; }
+	public string? m_Data { get; set; }
+}
+
+public class ModSerializedType
+{
+	public string? m_AssemblyName { get; set; }
+	public string? m_ClassName { get; set; }
+}
```
A few things:

- Making any class `public` requires a strong justification. These should be `internal`.
- Similarly, all instance classes should be `sealed` unless inheritance is intended.
- Separate classes go in separate files.

ds5678 commented on ModComponent/Utils/NameUtils.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

ds5678 commented on ModComponent/Utils/PlayerUtils.cs:
```diff
@@ -15,30 +15,25 @@ public static void UnfreezePlayer()
-		if (InterfaceManager.m_Panel_OptionsMenu?.m_State == null)
+		if (InterfaceManager.GetPanel<Panel_OptionsMenu>()?.State == null)
 		{
-			return PlayerGender.Unknown;
+			return Gender.Unknown;
 		}
 
-		if (GameManager.GetPlayerManagerComponent().m_VoicePersona != InterfaceManager.m_Panel_OptionsMenu.m_State.m_VoicePersona)
+		if (PlayerManager.m_VoicePersona == VoicePersona.Female)
 		{
-			return PlayerGender.Unknown;
-		}
-
-		if (InterfaceManager.m_Panel_OptionsMenu.m_State.m_VoicePersona == VoicePersona.Female)
-		{
-			return PlayerGender.Female;
+			return Gender.Female;
 		}
 		else
 		{
-			return PlayerGender.Male;
+			return Gender.Male;
 		}
 	}
```
If this isn't used in ModComponent, remove it.

ds5678 commented on ModComponent/Utils/UIUtils.cs:
```diff
@@ -1,5 +1,5 @@
-﻿extern alias Hinterland;
-using Hinterland;
+﻿using Il2Cpp;
+
```
Remove

dommrogers committed:
PR Changes #2 - See Comments

dommrogers committed:
remove name ref from BedMapper

dommrogers committed:
restored SetTexture _dmg_texture

dommrogers committed:
restore LoadJsonLocalization

and use same BOM check as LocalizationUtilities ( new method ReadToJsonString() )

ds5678 commented on ModComponent/Mapper/ZipFileLoader.cs:
```diff
@@ -87,6 +86,18 @@ private static string ReadToString(MemoryStream memoryStream)
+		}
+		return Encoding.UTF8.GetString(new ReadOnlySpan<byte>(bytes, index, bytes.Length - index));
+	}
+
```
Remove

ds5678 reviewed (changes requested):
These 3 commits look mostly good. I left my thoughts on the UTF8 bom issue. I'll do a full review soon.
